packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
+    onEvent,
+    onFatalError,
+    eventSelectors,
+    hasFactorySource: sources.some(sourceIsFactory),
+    logFilterSources: sources.filter(sourceIsLog),
+    factorySources: sources.filter(sourceIsFactory),
   };
+};
 
+export const start = (realtimeSyncService: Service) => {
   /**
-   * Possible states of process
-   *
-   * 1) Nothing is being processed
-   * 2) Processing block, no follow up invocation
-   * 3) Processing block, follow up invocation
+   * The queue reacts to a new block. The four states are:
+   * 1) Block is the same as the one just processed, no-op.
+   * 2) Block is exactly one block ahead of the last processed,
+   *    handle this new block (happy path).
+   * 3) Block is more than one ahead of the last processed,
+   *    fetch all intermediate blocks and enqueue them again.
+   * 4) Block is behind the last processed. This is a sign that
+   *    a reorg has occurred.
    */
-  process = async () => {
-    if (this.isProcessingBlock) {
-      this.isProcessBlockQueued = true;
-      return;
-    }
-
-    this.isProcessingBlock = true;
-    this.isProcessBlockQueued = false;
-
-    for (let i = 0; i < 6; i++) {
-      try {
-        const block = await this._eth_getBlockByNumber("latest");
-        await this.handleNewBlock(block);
-        break;
-      } catch (error_) {
-        const error = error_ as Error;
-        if (this.isShuttingDown) return;
-
-        this.common.logger.warn({
+  const queue = createQueue({
+    browser: false,
+    concurrency: 1,
+    initialStart: true,
+    worker: async (newBlock: SyncBlock) => {
+      const latestLocalBlock = getLatestLocalBlock(realtimeSyncService);
+      const newBlockNumber = hexToNumber(newBlock.number);
+
+      // We already saw and handled this block. No-op.
+      if (latestLocalBlock.hash === newBlock.hash) {
+        realtimeSyncService.common.logger.trace({
           service: "realtime",
-          msg: `Realtime sync task failed (network=${
-            this.network.name
-          }, error=${`${error.name}: ${error.message}`})`,
-          network: this.network.name,
+          msg: `Already processed block at ${newBlockNumber} (network=${realtimeSyncService.network.name})`,
         });
 
-        if (i === 5) this.emit("fatal", error);
-        else {
-          const duration = 250 * 2 ** i;
-          await wait(duration);
-        }
+        return;
       }
-    }
 
-    this.isProcessingBlock = false;
+      for (let i = 0; i < 6; i++) {
+        try {
+          // Quickly check for a reorg by comparing block numbers. If the block
+          // number has not increased, a reorg must have occurred.
+          if (latestLocalBlock.number >= newBlockNumber) {
+            await handleReorg(realtimeSyncService, newBlock);
 
-    if (this.isProcessBlockQueued) {
-      this.isProcessBlockQueued = false;
-      await this.process();
-    } else {
-      this.emit("idle");
-    }
-  };
+            if (realtimeSyncService.isKilled) return;
```

Comment:
We need this in order to quickly exit on kill. We want to avoid enqueing anything after the queue has been paused, which would cause the queue.onIdle to never resolve.

Right now the kill function sets the isKilled state variable, pauses and clears the queue, and then waits for it to become idle (no pending or unstarted tasks)